File: webapp/records.js
/* Game records — JSON "graph-SGF" (docs/game-record-format.md), shared schema with
   tilinggo/records.py. Plus standard SGF export for square boards only (SGF needs a
   coordinate grid; other substrates have none). */
(function (root) {
  "use strict";
  const FORMAT = "euclidean-go-record", VERSION = 1;

  function buildRecord(key, board, moves, meta) {
    meta = meta || {};
    return {
      format: FORMAT, version: VERSION,
      board: { key, fingerprint: root.SHARE.fingerprint(board), nodes: board.n },
      komi: board.komi, rules: "area",
      players: meta.players || {}, date: meta.date || null, result: meta.result || null,
      moves: moves.map(m => m === board.pass ? "pass" : m),
    };
  }

  /* Validate a parsed record object; returns a SHARE-style {key, fp, moves} ready for
     SHARE.replay / the app's installGame. Throws SHARE.ShareError with a typed code. */
  function toParsed(rec, TG, BOARDS) {
    const E = root.SHARE.ShareError;
    if (!rec || rec.format !== FORMAT) throw new E("BAD_FORMAT", "not a Euclidean Go record file");
    if (rec.version !== VERSION) throw new E("BAD_VERSION", `unsupported record version ${rec.version}`);
    const b = rec.board || {};
    if (typeof b.key !== "string" || !BOARDS[b.key])
      throw new E("BAD_BOARD", `this site doesn't have board "${b.key}"`);
    const board = TG.makeBoard(BOARDS[b.key]);
    if (b.fingerprint && b.fingerprint !== root.SHARE.fingerprint(board))
      throw new E("BAD_BOARD", "record was made for a different build of this board");
    if (b.graph && Array.isArray(b.graph.edges)) {           // embedded graph must agree
      const emb = b.graph.edges.flat();
      if (emb.length !== board.edges.length || emb.some((v, i) => v !== board.edges[i]))
        throw new E("BAD_BOARD", "embedded graph does not match the named substrate");
    }
    const moves = (rec.moves || []).map(m => {
      if (m === "pass") return root.SHARE.PASS;
      if (!Number.isInteger(m)) throw new E("BAD_FORMAT", `malformed move ${JSON.stringify(m)}`);
      return m;
    });
    return { version: 1, key: b.key, fp: root.SHARE.fingerprint(board), moves };
  }

  /* SGF (FF[4], Chinese/area rules) for square boards only. Coordinates come from the baked
     node positions: col = round(x), row = round(y); SGF letters are 'a'+index. */
  function toSGF(key, board, moves, meta) {
    const cols = board.coords.map(c => Math.round(c[0])), rows = board.coords.map(c => Math.round(c[1]));
    const w = Math.max(...cols) + 1, h = Math.max(...rows) + 1;
    const sz = (w === h) ? `${w}` : `${w}:${h}`;
    const L = i => String.fromCharCode(97 + i);
    let out = `(;FF[4]GM[1]CA[UTF-8]AP[euclidean-go]SZ[${sz}]RU[Chinese]KM[${board.komi}]`;
    if (meta && meta.result) out += `RE[${meta.result}]`;
    moves.forEach((m, i) => {
      const tag = (i % 2 === 0) ? "B" : "W";
      out += `;${tag}[` + (m === board.pass ? "" : L(cols[m]) + L(rows[m])) + "]";
    });
    return out + ")";
  }

  const REC = { FORMAT, VERSION, buildRecord, toParsed, toSGF };
  root.REC = REC;
  if (typeof module !== "undefined") module.exports = REC;
})(typeof globalThis !== "undefined" ? globalThis : this);

File: tilinggo/records.py
"""Game records — the JSON "graph-SGF" shared with the webapp (docs/game-record-format.md).

SGF assumes a coordinate grid, which only exists on the square board; Koala games are
sequences of *node indices* into a named board graph. A record stores the board key, a graph
fingerprint (and optionally the full edge list), metadata, and the move list. The same schema
is read/written by the webapp (webapp/records.js); cross-language identity is tested in
tests/test_records_cross.py.

Note an API asymmetry forced by the engine design: :class:`~tilinggo.rules.GoState` is an
immutable position and does not carry the move sequence that produced it, so ``save_record``
takes the move list explicitly; ``load_record`` replays the record and returns the final state.
"""

from __future__ import annotations

import json
from pathlib import Path

import numpy as np

from .rules import Board, GoState

FORMAT = "euclidean-go-record"
VERSION = 1
PASS = "pass"


class RecordError(ValueError):
    def __init__(self, code: str, message: str):
        super().__init__(message)
        self.code = code


def fingerprint(board: Board) -> str:
    """FNV-1a over node count + flattened edge list — byte-for-byte the webapp's algorithm
    (webapp/share.js), so records and links interoperate."""
    h = 0x811C9DC5

    def mix_int(v: int) -> None:
        nonlocal h
        for shift in (0, 8, 16, 24):
            h ^= (v >> shift) & 0xFF
            h = (h * 0x01000193) & 0xFFFFFFFF

    mix_int(board.n)
    for v in np.asarray(board.graph.edges, dtype=np.int64).ravel():
        mix_int(int(v))
    out, n = "", h
    while n:
        out = "0123456789abcdefghijklmnopqrstuvwxyz"[n % 36] + out
        n //= 36
    return (out or "0")[:6]


def _resolve_board(key: str, komi: float) -> Board:
    from .ui.server import _make_board  # the families×sizes catalogue

    return _make_board(key, komi=komi)


def save_record(path: str | Path, *, key: str, moves: list[int], board: Board | None = None,
                players: dict | None = None, result: str | None = None,
                date: str | None = None, embed_graph: bool = False) -> Path:
    """Write a game record. ``moves`` are node indices in play order (``board.pass_move`` or
    the string "pass" for passes)."""
    board = board if board is not None else _resolve_board(key, komi=5.5)
    norm = [PASS if (m == PASS or int(m) == board.pass_move) else int(m) for m in moves]
    rec = {
        "format": FORMAT,
        "version": VERSION,
        "board": {"key": key, "fingerprint": fingerprint(board), "nodes": board.n},
        "komi": board.komi,
        "rules": "area",
        "players": players or {},
        "date": date,
        "result": result,
        "moves": norm,
    }
    if embed_graph:
        rec["board"]["graph"] = {
            "edges": np.asarray(board.graph.edges, dtype=int).reshape(-1, 2).tolist()}
    p = Path(path)
    p.write_text(json.dumps(rec, indent=1))
    return p


def load_record(path: str | Path) -> GoState:
    """Parse + validate a record, replay it through the rules engine, return the final state."""
    try:
        rec = json.loads(Path(path).read_text())
    except (OSError, json.JSONDecodeError) as e:
        raise RecordError("BAD_FILE", f"not a readable JSON record: {e}") from e
    if rec.get("format") != FORMAT:
        raise RecordError("BAD_FORMAT", "not a euclidean-go-record file")
    if rec.get("version") != VERSION:
        raise RecordError("BAD_VERSION", f"unsupported record version {rec.get('version')}")
    b = rec.get("board") or {}
    key = b.get("key")
    if not isinstance(key, str):
        raise RecordError("BAD_FORMAT", "record has no board key")
    try:
        board = _resolve_board(key, komi=float(rec.get("komi", 5.5)))
    except Exception as e:
        raise RecordError("BAD_BOARD", f"unknown board {key!r}") from e
    if b.get("fingerprint") and b["fingerprint"] != fingerprint(board):
        raise RecordError("BAD_BOARD",
                          f"record was made for a different build of board {key!r}")
    if "graph" in b:  # embedded graph must agree with the compiled substrate
        emb = np.asarray(b["graph"].get("edges", []), dtype=int).reshape(-1, 2)
        ours = np.asarray(board.graph.edges, dtype=int).reshape(-1, 2)
        if emb.shape != ours.shape or not np.array_equal(emb, ours):
            raise RecordError("BAD_BOARD",
                              "embedded graph does not match the named substrate's compiled graph")

    state = board.new_game()
    for ply, m in enumerate(rec.get("moves", [])):
        node = board.pass_move if m == PASS else m
        if not (node == board.pass_move or (isinstance(node, int) and 0 <= node < board.n)):
            raise RecordError("BAD_MOVE", f"move {m!r} out of range at ply {ply}")
        if not state.legal_moves()[node]:
            raise RecordError("ILLEGAL_MOVE", f"move {m!r} is illegal at ply {ply}")
        state = state.play(node)
    return state
